### PR TITLE
[WIP] set the tokenizer vocab size

### PR DESCRIPTION
**Pull request by SaulLu**

### CODE REVIEW

File: megatron/mpu/layers.py [hunk 1]
```diff
@@ -217,6 +217,9 @@ def __init__(self, num_embeddings, embedding_dim,
 
 
     def forward(self, input_):
+        if torch.any(input_ >= self.num_embeddings):
+            raise ValueError("There is an input id in the input that is greater than the highest possible input id.")
+
```
Comment (SaulLu) [inline megatron/mpu/layers.py:222]: Before this check, with TP>1 the model was able to silently process input ids sequences with ids ouside of the admissible input ids.

File: megatron/arguments.py [hunk 2]
```diff
@@ -369,6 +369,11 @@ def _add_network_size_args(parser):
     group.add_argument('--make-vocab-size-divisible-by', type=int, default=128,
                        help='Pad the vocab size to be divisible by this value.'
                        'This is added for computational efficieny reasons.')
+    group.add_argument('--pad-vocab-size-to', type=int, default=None,
```
Comment (thomasw21) [inline megatron/arguments.py:372]:
```suggestion
    group.add_argument('--pad-embedding-size-to', type=int, default=None,
```

File: megatron/arguments.py [hunk 3]
```diff
@@ -369,6 +369,11 @@ def _add_network_size_args(parser):
...
+                       help='Pad the vocab size to be divisible by this value.'
+                       'Value of the size of the vocabulary of the tokenizer to reach. This value must be greater than'
```
Comment (thomasw21) [inline megatron/arguments.py:374]:
```suggestion
                       'Embedding size value, in terms of vocabulary size. . This value must be greater than'
```

File: megatron/arguments.py [hunk 4]
```diff
@@ -369,6 +369,11 @@ def _add_network_size_args(parser):
     group.add_argument('--make-vocab-size-divisible-by', type=int, default=128,
                        help='Pad the vocab size to be divisible by this value.'
                        'This is added for computational efficieny reasons.')
+    group.add_argument('--pad-vocab-size-to', type=int, default=None,
+                       help='Pad the vocab size to be divisible by this value.'
```
Comment (thomasw21) [inline megatron/arguments.py:373]:
```suggestion
                       help='Pad the embedding size to this value.'
```

File: megatron/arguments.py [hunk 5]
```diff
@@ -369,6 +369,11 @@ def _add_network_size_args(parser):
...
+                       'Value of the size of the vocabulary of the tokenizer to reach. This value must be greater than'
+                       ' the initial size of the tokenizer. If this argument is used the value of '
```
Comment (thomasw21) [inline megatron/arguments.py:375]:
```suggestion
                       ' the initial size of the tokenizer. This value also needs to be divisible by TP size. If this argument is used the value of '
```

File: megatron/mpu/layers.py [hunk 6]
```diff
@@ -217,6 +217,9 @@ def __init__(self, num_embeddings, embedding_dim,
 
 
     def forward(self, input_):
+        if torch.any(input_ >= self.num_embeddings):
+            raise ValueError("There is an input id in the input that is greater than the highest possible input id.")
```
Comment (thomasw21) [inline megatron/mpu/layers.py:221]:
```suggestion
            raise ValueError(f"There is an input id in the input that is greater than the highest possible input id.\nInput: {input_}\nnum_embeddings: {self.num_embeddings}")
```

File: megatron/mpu/layers.py [hunk 7]
```diff
@@ -225,7 +228,9 @@ def forward(self, input_):
             masked_input = input_.clone() - self.vocab_start_index
             masked_input[input_mask] = 0
         else:
+            # input_ is indeed in the range [0:self.vocab_end_index - self.vocab_start_index] thanks to the first check
```
Comment (thomasw21) [inline megatron/mpu/layers.py:231]: This is actually guaranteed via the `if else` statement, so you should put this on top of `if` or below (as from this point on this conditions are implicitely verified. ), just not inside.

File: megatron/tokenizer/tokenizer.py [hunk 8]
```diff
@@ -68,14 +68,22 @@ def build_tokenizer(args):
...
+            raise ValueError(
+                f"You asked to pad the vocabulary to {after} when the initial vocabulary size is "
+                f"{args.pad_vocab_size_to}. You can only pad to a higher value."
+            )
+    else:
+        # Pad vocab size so it is divisible by model parallel size and still having GPU friendly size.
+        multiple = args.make_vocab_size_divisible_by * \
+            args.tensor_model_parallel_size
+        while (after % multiple) != 0:
+            after += 1
```
Comment (thomasw21) [inline megatron/tokenizer/tokenizer.py:86]:
```suggestion
    if args.pad_vocab_size_to is not None:
        if args.pad_vocab_size_to  < orig_vocab_size:
            raise ValueError(
                f"You asked to pad the vocabulary to {args.pad_vocab_size_to} when the initial vocabulary size is "
                f"{orig_vocab_size}. You can only pad to a higher value."
            )
        after = args.pad_vocab_size_to
    else:
        # Pad vocab size so it is divisible by model parallel size and still having GPU friendly size.
        after = orig_vocab_size
        multiple = args.make_vocab_size_divisible_by * \
            args.tensor_model_parallel_size
        while (after % multiple) != 0:
            after += 1
```
You need to remove `after = orig_vocab_size`

Also though I know that you check for multiple of TP down the line, I think it'd make sense to have the check here.

Also I'd be in favor of not ignoring `make_vocab_divisible_by`. Just check that it works, if it doesn't then we raise an exception.

File: tests/test_tensor_parallel.py [hunk 9]
```diff
@@ -169,6 +166,7 @@ def test_alibi_tp(self):
         cp_dir = self.get_auto_remove_tmp_dir()
         
         command_args = self.get_default_args()
+        command_args["--pad-vocab-size-to"] = "50432" # This is equal to 128 * 394 which is above the len of gp2 vocabulary
```
Comment (thomasw21) [inline tests/test_tensor_parallel.py:169]: You should explain that we need TP=1 and TP=2 to have the same number of padded_vocab_size (you could try the smallest one closer to gpt2) ... Turns out this is the smallest multiple of 128 above gpt2 vocab right?

File: megatron/mpu/layers.py (hunk 1 above)
Comment (stas00) [inline megatron/mpu/layers.py:222]: killing the training at run time because the input is broken? That we can't support.

Additionally this error can't be acted upon since the operator would have no idea how to fix this as you're not including the input id and the same id - how can this get fixed?

If there is a need to validate data before the training it should happen separately from the training.

File: tests/test_tensor_parallel.py [hunk 10]
```diff
@@ -192,5 +190,91 @@ def test_alibi_tp(self):
...
+        command_args["--tensor-model-parallel-size"] = "2"
+
+        pool = Pool(2)
+        result = pool.map(MegDSTestTP.infer_model, [((0, 2, command_args, tokens, None, cp_dir)), ((1, 2, command_args, tokens, None, cp_dir))])
+        pool.close()
+        pool.join()
+        
+        output2, _ = result[0]
+
+        logging.getLogger().critical(output-output2)
```
Comment (thomasw21) [inline tests/test_tensor_parallel.py:225]: We don't need that do we?

File: megatron/mpu/layers.py (hunk 1 above)
Comment (stas00) [reply megatron/mpu/layers.py:222]: the forward should be as lean as possible and not need to check anything so that it could run as fast as possible.

File: tests/test_tensor_parallel.py [hunk 11]
```diff
@@ -192,5 +190,91 @@ def test_alibi_tp(self):
...
+
+        command_args["--tensor-model-parallel-size"] = "2"
+
+        pool = Pool(2)
+        with pytest.raises(Exception) as exc_info: 
+            _ = pool.map(MegDSTestTP.infer_model, [((0, 2, command_args, tokens, None, None)), ((1, 2, command_args, tokens, None, None))])
+        pool.close()
+        pool.join()
+
+        self.assertEqual(str(exc_info.value),"There is an input id in the input that is greater than the highest possible input id.")
```
Comment (thomasw21) [inline tests/test_tensor_parallel.py:258]: There's probably two exceptions no? We should catch both?

File: tests/test_tensor_parallel.py [hunk 12]
```diff
@@ -192,5 +190,91 @@ def test_alibi_tp(self):
...
+        command_args["--micro-batch-size"] = "2"
+        command_args["--tensor-model-parallel-size"] = "2"
+
+        pool = Pool(2)
+        with pytest.raises(Exception) as exc_info: 
+            _ = pool.map(MegDSTestTP.infer_model, [((0, 2, command_args, None, None, None)), ((1, 2, command_args, None, None, None))])
+        pool.close()
+        pool.join()
+
+        self.assertEqual(str(exc_info.value), "50433 is not divisible by 2")
```
Comment (thomasw21) [inline tests/test_tensor_parallel.py:273]: You sould put `--micro-batch-size` to 3 so that we're clear the 2 refers to TP.

File: megatron/mpu/layers.py (hunk 1 above)
Comment (thomasw21) [reply megatron/mpu/layers.py:222]: I don't agree, we need this sanity check so that we're not doing something wrong. Right now, model would silently bypass this issue if you use TP>2. IMO it should kill the training as we are doing something VERY bad.

File: tests/test_tensor_parallel.py (hunk 9 above)
Comment (stas00) [inline tests/test_tensor_parallel.py:169]: we are trying to keep the tests fast, you changed vocab size from 1024 to 50432, which is not fast. If it's necessary for the purpose of test, then it is ok, but please consider using small numbers.

File: megatron/arguments.py (hunk 2 above)
Comment (SaulLu) [reply megatron/arguments.py:372]: IMO, I liked the naming `vocab` because it emphasize the fact that we're really choosing our tokenizer's vocabulary size.

File: megatron/mpu/layers.py (hunk 7 above)
Comment (SaulLu) [reply megatron/mpu/layers.py:231]: But in the `if` statement it's explicit that is the case no?

File: megatron/mpu/layers.py (hunk 1 above)
Comment (stas00) [reply megatron/mpu/layers.py:222]: You don't do data sanity checks in `forward`. All sanity checks should ideally be done before at the dataloader level.

Which case are you guarding against - all of the data is completely borked or only some inputs are invalid?

Comment (stas00) [reply megatron/mpu/layers.py:222]: and practically, let's do a an imaginary scenario - you started the training and this assert happens 5 days in - what do you do? I fail to see how this assert is actionable.

File: tests/test_tensor_parallel.py (hunk 9 above)
Comment (SaulLu) [reply tests/test_tensor_parallel.py:169]: I need to use here a value greater than the initial vocabulary of the tokenizer. The initial tokenizer used is gpt2 which has a vocabulary of size ~50k tokens. 

We can work on creating a toy tokenizer - this requires time - for the test but I understood that we needed this feature ASAP.

File: megatron/mpu/layers.py (hunk 1 above)
Comment (SaulLu) [reply megatron/mpu/layers.py:222]: It will crash with the cross entropy loss no?

Comment (thomasw21) [reply megatron/mpu/layers.py:222]: We're trying to catch if some inputs are invalid. 

> and practically, let's do a an imaginary scenario - you started the training and this assert happens 5 days in - what do you do? I fail to see how this assert is actionable.

The same we would do if we ran in this exception using TP=1. We manually re-process the data, or try to remove that sample from the dataset ... We have this assert happen in TP=1 right now, so it makes sense not to have silent errors in TP=2.

File: tests/test_tensor_parallel.py (hunk 9 above)
Comment (thomasw21) [reply tests/test_tensor_parallel.py:169]: Yeah we can switch transformers for a smaller one if needed

Comment (SaulLu) [reply tests/test_tensor_parallel.py:169]: I change it to use `gpt2-tiny`

File: tests/test_tensor_parallel.py (hunk 12 above)
Comment (SaulLu) [reply tests/test_tensor_parallel.py:274]: That's a good point!
Done!

File: tests/test_tensor_parallel.py (hunk 9 above)
Comment (SaulLu) [reply tests/test_tensor_parallel.py:169]: > Turns out this is the smallest multiple of 128 above gpt2 vocab right?

Yes!

File: megatron/mpu/layers.py (hunk 1 above)
Comment (stas00) [reply megatron/mpu/layers.py:222]: how will you find this sample?

why force crash the program and force a huge down time until you figure it out if there was a mistake in a single input? 

why not move it into dataloader and just skip it and possibly warn about it in case it's not a single bad input but a lot of it?

Comment (SaulLu) [reply megatron/mpu/layers.py:222]: @Stas, just for information we also modify a little bit the error message in order to log also the input ids and the embedding size

File: tests/test_tensor_parallel.py (hunk 9 above)
Comment (SaulLu) [reply tests/test_tensor_parallel.py]: Done! Thanks for the tip

File: megatron/mpu/layers.py (hunk 1 above)
Comment (SaulLu) [reply megatron/mpu/layers.py:222]: Initially, this test was motivated by the fact that we wanted to be sure that the embedding matrix created with the tensor parallelism feature is the desired one. And we realized that contrary to TP=1, this type of error would pass without raising an error.

In my opinion, we can also add a test on the data side, but I think that this error is still relevant as we need to be sure that the embedding matrix is in agreement with the tokenizer we use. 

If you ever see another way to do it at the time of the creation of the model sharded architecture on the gpu's, it would be better but personally I don't see any place that would allow to do that since the "routing" of the input ids is done in the forward.

Comment (DanielHesslow) [reply megatron/mpu/layers.py:222]: Usually what happens if you index out of bounds is a completely undecipherable cuda error at a later point. It is definitely better to include this assert. All it does is improve the error message.

Another alternative I guess, if we really want to continue despite there being a bug is to send a warning and do `input_=min(input, self.num_embeddings)`, though I think I prefer the assert as we can't miss it and we should probably fix and restart it anyway if this happens.

File: tests/test_tensor_parallel.py (hunk 12 above)
Comment (SaulLu) [reply tests/test_tensor_parallel.py]: @thomasw21 , actually this test fail after this change - no error are raised. Do we really not this check? What is really a bad behavior if this is not a multiple of TP size?

**Comment by DanielHesslow** [comment]
Looks good to me. At some point we should probably factor out all of the pool/process launch stuff to a common place but that's for when things have calmed down a bit

File: megatron/mpu/layers.py (hunk 1 above)
Comment (stas00) [reply megatron/mpu/layers.py:222]: sorry, unfortunately I don't have the resources to dive right now into this as I have to finish lots of things before the launch, so I trust you will do the best thing you can and if things break during the training I will ping you and you will know what to do.

It's not great to have this sort of last minute change that hasn't been thoroughly tested but what to do.

Comment (stas00) [reply megatron/mpu/layers.py:222]: Thomas and I discussed this, he helped me to clear out an important misunderstanding and he will post an update.

Thank you for working on this, @SaulLu and @DanielHesslow - my apologies that I can't be involved at a deeper level at the moment.

Comment (thomasw21) [reply megatron/mpu/layers.py:222]: Okay talked to @stas00 :
 - This PR makes sense in the sense that the model should check for this. @DanielHesslow said it better, but if we get a out of bound index, we need to raise an exception
 - There's the issue of "if the training set is bad for X reason" what do we do? After this PR it will throw badly. So we need a strategy. Given that the number of root causes, @stas00 suggest to have a skip mechanism in the data loader, why? Because we're able to keep training, in case the issue "isn't that bad". In the worst case scenario, we kill the job and relaunch from scratch. But this is orthogonal to this PR IMO.

File: megatron/arguments.py (hunk 2 above)
Comment (thomasw21) [reply megatron/arguments.py:372]: actually works for me, I just didn't want to indicate we were modifying the tokenizer, but really just the embedding layer. But in Meg-DS they seem to use `padded_vocab_size` so your solution makes more sense.

File: megatron/mpu/layers.py (hunk 1 above)
Comment (stas00) [reply megatron/mpu/layers.py:222]: yes, so the idea is do a check on the dataloader level, and warn when skipping a bad sample. If it's a lot of samples but not the majority - we continue training while someone is fixing the data.

of course if the data is very broken then it'll be skipping them all and we can't train.

But the point is that - don't crash the training unless you have to.

Comment (stas00) [reply megatron/mpu/layers.py:222]: and additionally **we can** pre-check all our data outside of the training so that it doesn't have any bad samples.

This PR is mainly for future users, but in our training we should make sure it never hits this assert

File: megatron/mpu/layers.py (hunk 7 above)
Comment (thomasw21) [reply megatron/mpu/layers.py:231]: As you feel, I mean that line applies to both right? In the sense that both block make it so that `masked_input` is in hte correct range.

Comment (SaulLu) [reply megatron/mpu/layers.py:231]: I changed the wording a little bit to emphasize the fact that the values are in the good range thanks to the first check

**Review by thomasw21 — APPROVED** [review]
Thank you!
